src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to send sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+        return false;
+    }
+
+    // re-schedule event for after min 1hr if TTS is null or cannot be reached
+    ChipLogProgress(Zcl, "Emit sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+    return true;
+}
+
+static void utcTimeChanged(uint64_t utcTime)
+{
+    System::Clock::Seconds32 lastKnownGoodChipEpoch;
+    System::Clock::Microseconds64 realTime;
+    uint32_t utcTimetoChipEpoch;
+
+    Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+    System::SystemClock().GetClock_RealTime(realTime);
+    ChipLogError(Zcl, "UTCTime: %llu Last Known Good Time: %u Real Time: %llu", utcTime, lastKnownGoodChipEpoch.count(),
+                 realTime.count());
+    chip::UnixEpochToChipEpochTime((uint32_t)(utcTime / chip::kMicrosecondsPerSecond), utcTimetoChipEpoch);
+
+    if (utcTimetoChipEpoch >= lastKnownGoodChipEpoch.count()) // update Last Known Good Time if a more recent time is obtained
+    {
+        Server::GetInstance().GetFabricTable().SetLastKnownGoodChipEpochTime(System::Clock::Seconds32(utcTimetoChipEpoch));
+        System::SystemClock().SetClock_RealTime(System::Clock::Microseconds64(utcTime));
+
+        Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+        System::SystemClock().GetClock_RealTime(realTime);
+        ChipLogError(Zcl, " Last Known Good Time: %u Real Time: %llu", lastKnownGoodChipEpoch.count(), realTime.count());
+    }
+}
+
+bool emberAfTimeSynchronizationClusterSetUtcTimeCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetUtcTime::DecodableType & commandData)
+{
+    Optional<StatusCode> status = Optional<StatusCode>::Missing();
+    Status globalStatus         = Status::Success;
+
+    auto utcTime     = commandData.utcTime;
+    auto granularity = commandData.granularity;
+    auto timeSource  = commandData.timeSource;
+
+    TimeSynchronization::GranularityEnum currentGranularity;
+    Granularity::Get(commandPath.mEndpointId, &currentGranularity);
+
+    if (currentGranularity == TimeSynchronization::GranularityEnum::kNoTimeGranularity || granularity >= currentGranularity)
+    {
+        UTCTime::Set(commandPath.mEndpointId, utcTime);
+        utcTimeChanged(utcTime);
+        Granularity::Set(commandPath.mEndpointId, granularity);
+        if (timeSource.HasValue())
+        {
+            TimeSource::Set(commandPath.mEndpointId, timeSource.Value());
+        }
+        else
+        {
+            TimeSource::Set(commandPath.mEndpointId, TimeSynchronization::TimeSourceEnum::kAdmin);
+        }
+    }
+    else
+    {
+        status.Emplace(TimeSynchronization::StatusCode::kTimeNotAccepted);
```

Comment:
Why not just AddClusterSpecificFailure and return here, instead of carrying around the status variables?